core/src/saros/net/xmpp/contact/DiscoveryService.java

```diff
@@ -52,41 +54,52 @@ void connectionChanged(Connection connection) {
             return;
           }
           discoveryManager = ServiceDiscoveryManager.getInstanceFor(connection);
-          resourcesSarosSupport = new HashMap<>();
+          resourcesFeatureSupport = new HashMap<>();
         });
   }
 
   /**
-   * Query Feature Support for {@link SarosConstants#XMPP_FEATURE_NAMESPACE}.
+   * Query Feature Support.
    *
    * @param fullJid of contact
-   * @param resultCallback to call on positive result
+   * @param resultCallback to call on result
    */
-  void querySarosSupport(JID fullJid, Consumer<Boolean> resultCallback) {
+  void queryFeatureSupport(JID fullJid, Consumer<EnumSet<XMPPContact.Feature>> resultCallback) {
     discoveryExecutor.execute(
         () -> {
           String jid = fullJid.getRAW();
-          boolean supportsFeature;
-          if (resourcesSarosSupport.containsKey(jid))
-            supportsFeature = resourcesSarosSupport.get(jid);
-          else {
-            supportsFeature = runDiscovery(jid, SarosConstants.XMPP_FEATURE_NAMESPACE);
-            resourcesSarosSupport.put(jid, supportsFeature);
+          EnumSet<XMPPContact.Feature> features;
+          if (resourcesFeatureSupport.containsKey(jid)) {
+            features = resourcesFeatureSupport.get(jid);
+          } else {
+            features = runDiscovery(jid);
+            resourcesFeatureSupport.put(jid, features);
           }
 
-          if (supportsFeature) resultCallback.accept(supportsFeature);
+          resultCallback.accept(features);
         });
   }
 
-  private boolean runDiscovery(String fullJid, String feature) {
-    if (discoveryManager == null) return false;
+  private EnumSet<XMPPContact.Feature> runDiscovery(String fullJid) {
+    EnumSet<XMPPContact.Feature> features = EnumSet.noneOf(XMPPContact.Feature.class);
+
+    if (discoveryManager == null) return features;
     try {
       DiscoverInfo discoverInfo = discoveryManager.discoverInfo(fullJid);
-      return discoverInfo.containsFeature(feature);
+
+      for (Iterator<DiscoverInfo.Feature> it = discoverInfo.getFeatures(); it.hasNext(); ) {
+        String feature = it.next().getVar();
+        if (SarosConstants.XMPP_FEATURE_NAMESPACE.equals(feature)) {
```

Comment:
The whole logic does not match the description of this class. You hardwired what features are queried. You have no option to query anything else and you are moving business logic even more into the network layer.

Reply:
For clarification I added a link to `XMPPContact.Feature` in the class description.

Yes I hardwired the xmpp names to resolve features here, and added an abstraction for features with `XMPPContact.Feature` enum. My idea is that later this enum will be part of a generic Contact Class and its the job of the contact service to resolve them.

How would you like to change this?